### PR TITLE
[WIP]Electronic Billing with Xerox

### PR DESCRIPTION
**Pull request by hpatelserpentcs**
### Updated connector_acp
https://github.com/OCA/l10n-chile/issues/57

### CODE REVIEW

File: connector_acp/models/backend_acp.py [hunk 1]
```diff
@@ -19,6 +19,11 @@ class BackendAcp(models.Model):
     status = fields.Selection((('unconfirmed', 'Unconfirmed'),
                                ('confirmed', 'Confirmed')), string="Status",
                               default='unconfirmed')
+    type_of_connection = fields.Selection(
```
Comment (max3903) [inline connector_acp/models/backend_acp.py]:
```suggestion
    connection_type = fields.Selection(
```

File: connector_acp/models/backend_acp.py [hunk 2]
```diff
@@ -19,6 +19,11 @@ class BackendAcp(models.Model):
...
+        [('usb', 'USB'),
+         ('wifi', 'WIFI')],
```
Comment (dreispt) [inline connector_acp/models/backend_acp.py]: Make this an empty list for now.
If that error, then uase a single option `[('nd', 'Not defined')]`

File: connector_acp/models/etd_mixin.py [hunk 3]
```diff
@@ -67,44 +67,48 @@ def build_file(self, file):
...
+        file_list = []
+        for rec in self:
+            rec.set_jinja_env()
+
+            # Get the template
+            template = rec._env.from_string(base64.b64decode(file.template).
+                                             decode('utf-8'))
+            # Additional keywords used in the template
+            kwargs = rec.prepare_keywords()
+            kwargs.update({
```
Comment (dreispt) [inline connector_acp/models/etd_mixin.py]: I don't know it wasn't like that before: you can move this update into the `prepare_keywords()` method

**Comment by hpatelserpentcs** [comment]
@dreispt 

updated above comment and now coming traceback is occurred when validating xml schemas.

File: connector_acp/data/document_data_view.xml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+        <field name="name">Invoice</field>
+        <field name="model">account.invoice</field>
+    </record>
+
+    <record id="etd_account_ticket" model="etd.document">
+        <field name="name">Ticket</field>
+        <field name="model">account.invoice</field>
+    </record>
+
+</odoo>
```
Comment (dreispt) [inline connector_acp/data/document_data_view.xml:13]: This is a new data file, it needs to be added to the manifest.
Please add it in the `demo` key.

File: connector_acp/models/etd_mixin.py [hunk 5]
```diff
@@ -52,74 +54,95 @@ def get_etd_document(self):
...
+            'o': self,
+            'now': fields.datetime.now(),
+            'today': fields.datetime.today()
+        }
 
     def get_etd_filename(self, file):
         return '%s.%s' % (self.name or self.number, file.file_type)
 
+    def get_etd_directory(self, file, file_list, res_content):
+        filename = self.get_etd_filename(file)
```
Comment (dreispt) [inline connector_acp/models/etd_mixin.py]: Have this just `return` nothing.
Later modules will extend this to implement the Xerox specific logic.

File: connector_acp/models/etd_mixin.py [hunk 6]
```diff
@@ -52,74 +54,95 @@ def get_etd_document(self):
...
+        for file_dict in file_list:
+            if filename in file_dict:
+                file_dict['content'] += res_content
+
     def build_file(self, file):
-        """
```
Comment (dreispt) [inline connector_acp/models/etd_mixin.py:65]: Add optional argument files. 
If the filename generated for the current record exists in the files, the file template has the grouped flag set, then append content text to it.

File: connector_acp/models/etd_mixin.py [hunk 7]
```diff
@@ -52,74 +54,95 @@ def get_etd_document(self):
...
-                    base64.b64encode(res_file.filecontent.encode("utf-8")),
-                'datas_fname': res_file.filename,
-                'res_model': self._name,
-                'res_id': self.id})
-        return {'name': filename, 'content': res_content}
-
-    def build_files(self):
-        """
+        file_list = []
+        files = []
```
Comment (dreispt) [inline connector_acp/models/etd_mixin.py]:
```suggestion
        files = files or []
```

File: connector_acp/models/etd_mixin.py [hunk 8]
```diff
@@ -52,74 +54,95 @@ def get_etd_document(self):
...
+            # Get the template
+            template = rec._env.from_string(
+                base64.b64decode(file.template).decode('utf-8'))
+            # Additional keywords used in the template
+            kwargs = rec.prepare_keywords()
+            filename = rec.get_etd_filename(file)
+            # Render the file
+            res_file = rec.File_details(filename, template.render(kwargs))
+            res_content = str.encode(res_file.filecontent)
+            if file.file_type == 'txt':
```
Comment (dreispt) [inline connector_acp/models/etd_mixin.py:98]:
```suggestion
            if file.grouped and filename in [f['name'] for f in files]:
                file_rec = [f for f in files if f['name'] == filename]
                if file_rec:
                    file_rec['content'] += res_content 
```

**Review by dreispt — CHANGES_REQUESTED** [review]
Some changes to make.

**Pull request closed without merging**